haystack/nodes/base.py

```diff
@@ -80,6 +80,13 @@ def __init_subclass__(cls, **kwargs):
                 "See https://haystack.deepset.ai/pipeline_nodes/custom-nodes for more information."
             )
 
+        # If a class has no explicit __init__(), if will fail in exportable_to_yaml() and fail to work with get_config.
+        # This step makes sure that such check will be passed. See exportable_to_yaml() for details
+        # https://github.com/deepset-ai/haystack/issues/3291
+        name_components = cls.__init__.__qualname__.split(".")
+        if f"{name_components[-2]}.{name_components[-1]}" != f"{cls.__name__}.{cls.__init__.__name__}":
+            cls.__init__.__qualname__ = f"{cls.__name__}.{cls.__init__.__name__}"
```

Comment:
Just to understand that correctly:
if the if-condition is true, we know that `cls.__init__` is the `__init__` of the subclass?
And if so, is `cls.__init__` then a copy of the subclass's `__init__` or is it actually the same instance? If it were the same instance, changing its `__qualname__` would be dangerous.
In general we should avoid changing dunder-attributes of functions.

I'm not sure if it really works, but an easier, less dangerous solution might be to adjust the check in line 40 to
```python
if self.__class__.__init__ == init_func:
```

But maybe I'm missing something. WDYT?

Reply:
So:

1. Changing line 40 with `if self.__class__.__init__ == init_func:` makes `test_get_config_custom_node_with_no_explicit_init` fail.

2. However, my change does not fix the case where there are two subclasses with no `__init__` (I added a test called `test_get_config_many_custom_nodes_with_no_explicit_init`)

I'm going to ask for another review once I have some idea.